File: JavaScript/프로그래머스/2/42586. 기능개발/기능개발.js
class Queue{
    items = [];
    front = 0;
    rear = 0;
    size() {
        return (this.rear - this.front);
    }
    push(item){
        this.items.push(item);
        this.rear++;
    }
    getFront() {
        return this.items[this.front]
    }
    back(){
        return this.items[this.rear - 1];
    }
    upgrade() {
        this.items = [];
        this.front = 0;
        this.rear = 0;
    }
    isEmpty() {
        return this.front === this.rear;
    }
}

function solution(progresses, speeds) {
    var answer = [];
    const queue = new Queue();
    for (const [key, progress] of progresses.entries()){
        let calc = Math.ceil((100 - progress) / speeds[key]);
        if (queue.isEmpty()){
            queue.push(calc);
        } else {
            if (queue.getFront() >= calc){
                queue.push(calc);
            } else {
                answer.push(queue.size());
                queue.upgrade();
                queue.push(calc);
            }
        }
    }
    if (!queue.isEmpty()){
        answer.push(queue.size())
    }
    return answer;
}

File: C++/프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    queue<int> q;
    
    for (int i = 0; i < progresses.size(); i++){
        if ((100 - progresses[i]) % speeds[i] == 0){
            q.push((100 - progresses[i]) / speeds[i]);
        } else {
            q.push((100 - progresses[i]) / speeds[i] + 1);
        }
    }
    int temp;
    while (!q.empty()){
        temp = q.front();
        int cnt = 0;
        while (!q.empty() && q.front() <= temp){
            q.pop();
            cnt++;
        }
        answer.push_back(cnt);
    }
    
    return answer;
}